src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -1298,32 +1311,57 @@ private void OnHeaderCore(int? index, bool indexedValue, ReadOnlySpan<byte> name
             throw new Http2ConnectionErrorException(CoreStrings.BadRequest_HeadersExceedMaxTotalSize, Http2ErrorCode.PROTOCOL_ERROR);
         }
 
-        if (index != null)
-        {
-            ValidateStaticHeader(index.Value, value);
-        }
-        else
-        {
-            ValidateHeader(name, value);
-        }
-
         try
         {
             if (_requestHeaderParsingState == RequestHeaderParsingState.Trailers)
             {
+                UpdateHeaderParsingState(value, GetPseudoHeaderField(name));
+                ValidateHeaderContent(name, value);
+
                 _currentHeadersStream.OnTrailer(name, value);
             }
             else
             {
                 // Throws BadRequest for header count limit breaches.
                 // Throws InvalidOperation for bad encoding.
-                if (index != null)
+                switch (headerType)
                 {
-                    _currentHeadersStream.OnHeader(index.Value, indexedValue, name, value);
-                }
-                else
-                {
-                    _currentHeadersStream.OnHeader(name, value, checkForNewlineChars : true);
+                    case HeaderType.Static:
+                        ValidateStaticHeader(index.GetValueOrDefault(), value);
+
+                        _currentHeadersStream.OnHeader(index.GetValueOrDefault(), indexedValue: true, name, value);
+                        break;
+                    case HeaderType.StaticAndValue:
+                        ValidateStaticHeader(index.GetValueOrDefault(), value);
+
+                        _currentHeadersStream.OnHeader(index.GetValueOrDefault(), indexedValue: false, name, value);
+                        break;
+                    case HeaderType.Dynamic:
+                        // Clients will normally send pseudo headers as an indexed header.
+                        // Because pseudo headers can still be sent by name we need to check for them.
+                        UpdateHeaderParsingState(value, GetPseudoHeaderField(name));
+
+                        if (index != null)
+                        {
+                            // It is faster to set a header using a static table index than a name.
```

Comment:
"static table" Isn't this the dynamic table index?

Reply:
No, this is the static index of the dynamic table entry.

e.g. :path=/MyCustomPath is added to the dynamic table. The static index here is :path's value.